lms/djangoapps/courseware/views/views.py

```diff
@@ -1824,6 +1825,61 @@ def render_xblock(request, usage_key_string, check_if_enrolled=True):
         return render_to_response('courseware/courseware-chromeless.html', context)
 
 
+@require_http_methods(["GET"])
+@ensure_valid_usage_key
+@xframe_options_exempt
+@transaction.non_atomic_requests
+def render_public_video_xblock(request, usage_key_string):
+    """
+    Returns an HttpResponse with HTML content for the Video xBlock with the given usage_key.
+    The returned HTML is a chromeless rendering of the Video xBlock (excluding content of the containing courseware).
+    """
+    view = 'public_view'
+
+    usage_key = UsageKey.from_string(usage_key_string)
+    usage_key = usage_key.replace(course_key=modulestore().fill_in_run(usage_key.course_key))
+    course_key = usage_key.course_key
+
+    # usage key block type must be `video` else raise 404
+    if usage_key.block_type != 'video':
+        raise Http404("Video not found.")
+
+    with modulestore().bulk_operations(course_key):
+        try:
+            course = get_course_by_id(course_key, 0)
+        except CourseAccessRedirect:
+            raise Http404("Course not found.")  # lint-amnesty, pylint: disable=raise-missing-from
+
+        block, _ = get_module_by_usage_id(
+            request,
+            str(course_key),
+            str(usage_key),
+            disable_staff_debug_info=True,
+            course=course,
+            will_recheck_access=False
+        )
+
+        # video must be public (`Public Access` field set to True) by course author in studio in video advanced settings
+        if not block.public_access:
+            raise Http404("Video not found.")
+
+        fragment = block.render(view, context={})
+
+        context = {
+            'fragment': fragment,
+            'course': course,
+            'disable_accordion': False,
+            'allow_iframing': True,
+            'disable_header': False,
+            'disable_footer': False,
+            'disable_window_wrap': True,
+            'edx_notes_enabled': False,
+            'is_learning_mfe': True,
+            'is_mobile_app': False,
+        }
+        return render_to_response('courseware/courseware-chromeless.html', context)
```

Comment:
@mikix 
* The whole purpose of this new endpoint was to just render `videos` only.
* I will look if we can move remove this new endpoint and use `VideoBlock` to access videos publicly?
* Sorry, I am not able to understand this? Can you please explain more?

Reply:
1. Sure, the problem you are solving right now is videos. But put another way, you are just solving "allow public access to an xblock" and the only xblock you happen to care about now is videos.

I mean, there's a reason we don't have 15 different view endpoints, one for each xblock type. The whole system is very generic and delegated. That's why I was asking if it wouldn't be reasonable to do this more generically. Like, maybe in two years, we want to allow public access to another xblock type. I dunno.

But anyway, that involves more thinking and more work, like probably moving the `public_access` field to the base xblock. I was not trying to derail you. Just saying that a video-specific endpoint that then turns around and just uses generic xblock rendering code feels a bit like a mismatch and a lost opportunity for future extensibility. But probably not worth rewriting much. This does not prevent us from making a generic version in the future, which would be a graver sin.

2. Again, I didn't want to derail you. Just thought it might be cleaner to keep video-specific code in the video-specific class, but I also see the benefit of keeping the two xblock render endpoints next to each other. :shrug: Just me thinking out loud, and not a call to action necessarily on your part.

3. If you search for `public_view` in edx-platform, you'll see lots of references and comments like:

```
# This is the view that will be rendered to display the XBlock in the LMS for unenrolled learners.
# Implementations of this view should assume that a user and user data are not available.
```

Which sounds in the same sphere as what you are doing. `VideoBlock` has a defined `public_view` implementation (which just redirects to the `student_view` version for modern xblocks). I don't know anything about this feature, honestly. So maybe it's not fit for your purpose at all. But it just feels very close to what you're doing, and maybe could share implementation details. Probably not. But maybe it would be worth adding a comment in your view about how it differs from this potentially-confusingly-similar `public_view` feature.

We just have so many similar-but-different features that my hairs go up when I see us maybe doing that again. Most recently I had to deal with some entrance exam vs prerequisite nonsense and they are both very similar and probably _could_ have shared an implementation, but they did not. So I'm just a little sensitive to that right now  :)

Reply:
@mikix Bundle of Thanks for your valuable feedback. Your concerns are valid. I also have these concerns but unable to find a way to resolve.

I think, If somehow we merge the changes made in new endpoint(`render_public_video_xblock`) into the existing endpoint(`render_xblock`) then all the concerns will be resolved. 
a) We will be using the existing code. 
b) We don't need to maintain a new endpoint that is a subset of the existing endpoint .

I have tried initially but the below code in `render_xblock` endpoint is a hurdle
```python
try:
    course = get_course_with_access(request.user, 'load', course_key, check_if_enrolled=check_if_enrolled)
except CourseAccessRedirect:
    raise Http404("Course not found.") 
```

The above code will only give the course if
1. User is authenticated, OR
2. Whole course is public - please see [Public Course Content Access](https://openedx.atlassian.net/wiki/spaces/ENGAGE/pages/3184459792/Public+Course+Content+Access)

But in my case, we need course even if all the above conditions fail.

Also `render_xblock` endpoint is being used to render every single XBlock so it feels uncomfortable to add some specific use case changes in a generic code. That is my understanding. But there may be a way to merge the two endpoints.

Also your feedback is more than welcome. It is always good to have feedback to implement a great solution. Waiting for more feedback from your side.

Reply:
- Do we suspect that `public_view` is just there to support the public-course feature? Which in my memory is only used for exactly one covid-related course. We wrote it a million years ago, so we must have had a purpose to it, but it's very niche. I don't think _any_ courses use the public-outline setting.

- I share your worries about how to circumvent the course lookup in `render_xblock`. The module lookup code wants a course, but you don't know if you can skip course access controls until you load the module and can check `public_access`.

- If we were feeling ambitious, a path forward might be "load the item descriptor directly first and check if it's public. If so, render with public view. If not, load the course with access checks and hydrate the already-loaded descriptor into a module." The "check if it's public" bit would be video-specific if the `public_access` field stayed on the video module, but could be guarded with a `hasattr('public_access')` or a `type == 'video'` check. Or the field could be moved to the base xblock (though we don't really have any plans for anything that's not a video, yeah?)

- But I totally understand hesitancy around refactoring this section of sensitive code. I think your original version is a fine low-touch, low-risk approach.